### PR TITLE
Clang Tidy to collision_detection and collision_detection_fcl

### PR DESCRIPTION
**Pull request by BryceStevenWilley**
### Description

As per #28, I setup and documented how to run clang-tidy on the codebase, and tested it on the collision_detection and collision_detection_fcl folders. I chose these two because they were small, so I could get feedback on what to change before applying to the entire repo.

The clang-tidy config file is based on https://github.com/davetcoleman/rviz_visual_tools/blob/kinetic-devel/.clang-tidy and adds some more checks to the code.

### Checklist
- [x] **Required**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [x] Extended the tutorials / documentation, if necessary [reference](http://moveit.ros.org/documentation/contributing/)
- [x] Include a screenshot if changing a GUI
- [x] Optional: Created tests, which fail without this PR [reference](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/tests.html)
- [ ] Optional: Decide if this should be cherry-picked to other current ROS branches (Indigo, Jade, Kinetic)

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: moveit_core/collision_detection/src/allvalid/collision_robot_allvalid.cpp [hunk 1]
```diff
@@ -51,7 +51,7 @@ void collision_detection::CollisionRobotAllValid::checkSelfCollision(const Colli
 {
   res.collision = false;
   if (req.verbose)
-    logInform("Using AllValid collision detection. No collision checking is performed.");
+    CONSOLE_BRIDGE_logInform("Using AllValid collision detection. No collision checking is performed.");
```
Comment (davetcoleman) [inline moveit_core/collision_detection/src/allvalid/collision_robot_allvalid.cpp]: why is this change necessary? we weren't getting deprecated / compatibility warnings were we? I see [this file](http://docs.ros.org/api/class_loader/html/console__bridge__compatibility_8h.html) but would like to understand the backstory. This change reduces code readability.

File: moveit_core/README.md [hunk 2]
```diff
@@ -8,3 +8,16 @@ This repository includes core libraries used by MoveIt:
...
+Install clang-tidy with 
+`sudo apt-get install clang-tidy`
+
+To add it to a new package, add `set(CMAKE_EXPORT_COMPILE_COMMANDS ON)` to `CMakeLists.txt`.
+
+Run it on a specific folder, such as `collision_detection`, with
+```
+run-clang-tidy-3.8.py -fix -p=/home/brycew/Desktop/moveit_ws/build/moveit_core/  collision_detection
+```
+
```
Comment (davetcoleman) [inline moveit_core/README.md]: Thanks for writing this up! I believe this documentation belongs on [this page](http://moveit.ros.org/documentation/contributing/code/) instead

File: moveit_core/collision_detection/src/allvalid/collision_robot_allvalid.cpp (hunk 1 above)
Comment (BryceStevenWilley) [reply moveit_core/collision_detection/src/allvalid/collision_robot_allvalid.cpp]: We were getting deprecated warnings, actually. The exact warning that clang gives is this:
```
/home/brycew/Desktop/moveit_ws/src/moveit/moveit_core/collision_detection/src/allvalid/collision_robot_allvalid.cpp:54:5: warning: 'log_deprecated' is deprecated [clang-diagnostic-deprecated-declarations]
    logInform("Using AllValid collision detection. No collision checking is performed.");
    ^
/usr/include/console_bridge/console.h:96:19: note: expanded from macro 'logInform'
  console_bridge::log_deprecated(__FILE__, __LINE__, console_bridge::CONSOLE_BRIDGE_LOG_INFO,  fmt, ##__VA_ARGS__)
                  ^
/usr/include/console_bridge/console.h:202:54: note: 'log_deprecated' has been explicitly marked deprecated here
CONSOLE_BRIDGE_DEPRECATED CONSOLE_BRIDGE_DLLAPI void log_deprecated(const char *file, int line, LogLevel level, const char* m, ...);
```

And looks like [this file](https://github.com/ros/console_bridge/blob/master/include/console_bridge/console.h#L81) does deprecate the current version we're using unfortunately. 

And from [this pull request](https://github.com/ros/console_bridge/pull/48/files), it looks like `logInform` et al. are about to be removed soon.

File: moveit_core/README.md (hunk 2 above)
Comment (BryceStevenWilley) [reply moveit_core/README.md]: Thanks! I had a feeling there was a better place, but couldn't find where.
This has been moved to https://github.com/ros-planning/moveit.ros.org/pull/146

**Comment by davetcoleman** [comment]
Just need one more reviewer

**Comment by davetcoleman** [comment]
Actually we should probably merge https://github.com/ros-planning/moveit/pull/662 otherwise there will undoubtably be major conflicts.

**Comment by BryceStevenWilley** [comment]
Merged! Was surprisingly easy, only 3 or 4 clear-cut conflicts.

**Comment by BryceStevenWilley** [comment]
That sounds good @rhaschke, went ahead and reverted that merge.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h [hunk 3]
```diff
@@ -59,7 +59,7 @@ class CollisionRobot
    *  @param padding The padding to use for all objects/links on the robot
    *  @scale scale A common scaling to use for all objects/links on the robot
    */
-  CollisionRobot(const robot_model::RobotModelConstPtr& model, double padding = 0.0, double scale = 1.0);
+  CollisionRobot(robot_model::RobotModelConstPtr model, double padding = 0.0, double scale = 1.0);
```
Comment (rhaschke) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h]: Replacing const-ref by value argument is a suggestion of clang tidy??
This shouldn't be replaced.

File: moveit_core/collision_detection/src/collision_robot.cpp [hunk 4]
```diff
@@ -145,12 +144,12 @@ double collision_detection::CollisionRobot::getLinkPadding(const std::string& li
 void collision_detection::CollisionRobot::setLinkPadding(const std::map<std::string, double>& padding)
 {
   std::vector<std::string> u;
-  for (std::map<std::string, double>::const_iterator it = padding.begin(); it != padding.end(); ++it)
+  for (const auto& it : padding)
```
Comment (rhaschke) [inline moveit_core/collision_detection/src/collision_robot.cpp]: here the loop variable should be renamed to e.g. `pair` (`it` is misleading).

File: moveit_core/collision_detection/src/collision_robot.cpp [hunk 5]
```diff
@@ -184,12 +183,12 @@ double collision_detection::CollisionRobot::getLinkScale(const std::string& link
 void collision_detection::CollisionRobot::setLinkScale(const std::map<std::string, double>& scale)
 {
   std::vector<std::string> u;
-  for (std::map<std::string, double>::const_iterator it = scale.begin(); it != scale.end(); ++it)
+  for (const auto& it : scale)
```
Comment (rhaschke) [inline moveit_core/collision_detection/src/collision_robot.cpp]: dito: `it` -> `pair`

File: moveit_core/collision_detection/src/collision_robot.cpp [hunk 6]
```diff
@@ -203,12 +202,12 @@ const std::map<std::string, double>& collision_detection::CollisionRobot::getLin
 void collision_detection::CollisionRobot::setPadding(const std::vector<moveit_msgs::LinkPadding>& padding)
 {
   std::vector<std::string> u;
-  for (std::size_t i = 0; i < padding.size(); ++i)
+  for (const auto& i : padding)
```
Comment (rhaschke) [inline moveit_core/collision_detection/src/collision_robot.cpp]: `i` -> `p`

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h (hunk 3 above)
Comment (BryceStevenWilley) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h]: The reasons for the replacement are on this page: https://clang.llvm.org/extra/clang-tidy/checks/modernize-pass-by-value.html. Looks like it only does so for constructors at the moment.

Easy enough to turn off though, I will do so.

Comment (rhaschke) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h]: Explanation makes sense in general. However, for shared ptrs, move-construction doesn't make sense. Usually you want to keep the pointer afterwards. So, no need to turn this feature off, but asking for manual validation of all changes (and reverting unnecessary ones).

**Comment by rhaschke** [comment]
Closing in favor of #731, which is a rebase of this PR completely removing the reverted merge.
